demos/common/hfsm/az_iot_hfsm_sync_adapter.c

```diff
@@ -0,0 +1,309 @@
+
+/* Include logging header files and define logging macros in the following order:
+ * 1. Include the header file "logging_levels.h".
+ * 2. Define the LIBRARY_LOG_NAME and LIBRARY_LOG_LEVEL macros depending on
+ * the logging configuration for DEMO.
+ * 3. Include the header file "logging_stack.h", if logging is enabled for DEMO.
+ */
+
+#include "logging_levels.h"
+
+/* Logging configuration for the Demo. */
+#ifndef LIBRARY_LOG_NAME
+    #define LIBRARY_LOG_NAME    "IoT_HFSM_SYNC"
+#endif
+
+#ifndef LIBRARY_LOG_LEVEL
+    #define LIBRARY_LOG_LEVEL    LOG_INFO
+#endif
+
+/* 
+ * The function prints to the console before the network is connected;
+ * then a UDP port after the network has connected. */
+extern void vLoggingPrintf( const char * pcFormatString,
+                            ... );
+
+/* Map the SdkLog macro to the logging function to enable logging */
+#ifndef SdkLog
+    #define SdkLog( message )    vLoggingPrintf message
+#endif
+
+#include "logging_stack.h"
+/**************************************************/
+
+// The retry Hierarchical Finite State Machine object.
+// A single Provisioning + Hub client is supported when syncrhonous mode is used.
+static az_iot_hfsm_type iot_hfsm;
+
+// HFSM adapters for Provisioning and Hub operations.
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+static az_hfsm provisioning_hfsm;
+#endif
+static az_hfsm hub_hfsm;
+
+// The two HFSM objects share the same implementation. Only one is active at any time.
+#ifdef AZ_IOT_HFSM_PROVISIONING_ENABLED
+static az_hfsm* active_hfsm = &provisioning_hfsm;
+#else
+static az_hfsm* active_hfsm = &hub_hfsm;
+#endif
+
+typedef enum
+{
+  AZ_IOT_HFSM_SYNC_DO_WORK = AZ_HFSM_EVENT(1),
+  AZ_IOT_HFSM_SYNC_DO_DELAY = AZ_HFSM_EVENT(2)
+} az_iot_hfsm_sync_event_type;
+
+static const az_hfsm_event az_hfsm_event_do_work = { AZ_IOT_HFSM_SYNC_DO_WORK, NULL };
+static const az_hfsm_event az_hfsm_event_do_delay = { AZ_IOT_HFSM_SYNC_DO_DELAY, NULL };
+
+static az_hfsm_event az_hfsm_sync_event_error = { AZ_IOT_ERROR, NULL };
+
+static int32_t delay_milliseconds;
+
+static int32_t azure_iot_sync(az_hfsm* me, az_hfsm_event event);
+static int32_t idle(az_hfsm* me, az_hfsm_event event);
+static int32_t running(az_hfsm* me, az_hfsm_event event);
+static int32_t timeout(az_hfsm* me, az_hfsm_event event);
+
+static az_hfsm_state_handler provisioning_and_hub_get_parent(az_hfsm_state_handler child_state)
+{
+  az_hfsm_state_handler parent_state;
+
+  if ((child_state == azure_iot_sync))
+  {
+    parent_state = NULL;
+  }
+  else 
+  {
+    parent_state = azure_iot_sync;
+  }
+
+  return parent_state;
+}
+
+static int32_t azure_iot_sync(az_hfsm* me, az_hfsm_event event)
+{
+  switch ((int32_t)event.type)
+  {
+    case AZ_HFSM_ENTRY:
+    case AZ_HFSM_EXIT:
+      break;
+    
+    default:
+      // Control should never reach this azure_iot_sync state.
+      LogInfo( ("azure_iot_sync: PANIC!") );
+      az_iot_hfsm_pal_critical_error(me);
+  }
+
+  return 0;
```

Comment:
why even have a return value? Future proof?

Reply:
No: this is required for hierarchical execution. Since you're asking I see that there is a gap in semantic of the value 0 which is really `AZ_HFSM_RET_HANDLED` - I'll add this.

Inner states may return AZ_HFSM_RET_HANDLE_BY_SUPERSTATE if they weren't able to handle the event.